### PR TITLE
Cosmology equality

### PR DESCRIPTION
**Pull request by nstarman**
Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

```python
>>> from astropy.cosmology import Planck13, Planck18

>>> Planck13 == Planck18
False

>>> Planck18 == Planck18
True
```

Blocking: #11559

### CODE REVIEW

File: astropy/cosmology/core.py [hunk 1]
```diff
@@ -198,6 +198,37 @@ def clone(self, *, meta=None, **kwargs):
...
+        """
+        if not isinstance(other, Cosmology):
+            return False
+
+        sias, oias = self._init_arguments, other._init_arguments
+
+        # check if the cosmologies have identical signatures.
+        # this protects against one cosmology having a superset of input
+        # parameters to another cosmology.
+        if set(sias.keys()).difference(oias.keys()):
```
Comment (mhvk) [inline astropy/cosmology/core.py:225]: This is not correct: if `oias` has more keys, the result will be empty; just do `if sias.keys() != oias.keys()` (this is not sensitive to order).

Probably good to test this case explicitly!

p.s. If `meta` is ignored below, should it be ignored here too? If so,
`if (sios.kays() ^ oias.keys()) - {'meta'}:` (this `set` symmetric difference operation works on dict keys, it seems, though not sure how long this has been the case)

File: astropy/cosmology/core.py [hunk 2]
```diff
@@ -118,14 +119,13 @@ def __init_subclass__(cls):
         cls._init_signature = sig  # store (immutable) initialization signature
 
     def __new__(cls, *args, **kwargs):
-        # bundle initialization argument
+        self = super().__new__(cls)
```
Comment (mhvk) [inline astropy/cosmology/core.py:122]: The change in `__new__` seems unrelated and makes things less clear, so I'd not do it.

File: astropy/cosmology/tests/test_cosmology.py [hunk 3]
```diff
@@ -305,6 +305,23 @@ def test_clone():
...
+    # Equality
+    assert Planck18 == Planck18
+    assert Planck13 != Planck18
+
+    # just wrong
+    assert Planck18 != 2
+    assert 2 != Planck18
+
+    # mismatched signatures
+    assert core.w0waCDM(**Planck18._init_arguments, Ode0=0.6) != Planck18
```
Comment (mhvk) [inline astropy/cosmology/tests/test_cosmology.py:319]: So, here also add `assert Planck18 != ...`

**Review by mhvk — COMMENTED** [review]
Good idea! See inline comments.

File: astropy/cosmology/core.py (hunk 1 above)
Comment (nstarman) [reply astropy/cosmology/core.py:225]: Thanks for the suggestion. I had assumed that since dictionaries are now ordered, so too were the keys. Good to know.

I guess it's possible for a user-defined Cosmology to not have "meta" in the constructor signature, so explicitly excluding it will avoid that rare edge case. I'll check the python version for symmetric difference.

File: astropy/cosmology/core.py (hunk 2 above)
Comment (nstarman) [reply astropy/cosmology/core.py:122]: I'll move it to another PR. The MappingProxyType protection on `_init_arguments` is good for ensuring it is immutable.

**Review by mhvk — APPROVED** [review]
Looks all OK now!

**Comment by nstarman** [comment]
Thanks for the key comparison technique!

**Comment by pllim** [comment]
Gentle reminder to milestone before merge. The milestone check bot is broken, so don't be fooled by the green checkmark. I have added 5.0 milestone for this one. FYI.
